File: PYTHON  PROJECTS/temperature_converter.py
# temperature_converter.py
# A robust temperature conversion tool that converts between Celsius, Fahrenheit, and Kelvin.
# Designed for Hacktoberfest contributions with modular design, error handling, and user-friendly interface.

# Conversion Functions
def celsius_to_fahrenheit(celsius):
    # Converts temperature from Celsius to Fahrenheit using the formula: (°C * 9/5) + 32
    return (celsius * 9/5) + 32

def fahrenheit_to_celsius(fahrenheit):
    # Converts temperature from Fahrenheit to Celsius using the formula: (°F - 32) * 5/9
    return (fahrenheit - 32) * 5/9

def celsius_to_kelvin(celsius):
    # Converts temperature from Celsius to Kelvin by adding 273.15
    return celsius + 273.15

def kelvin_to_celsius(kelvin):
    # Converts temperature from Kelvin to Celsius by subtracting 273.15
    # Validates that the input is not below absolute zero (0K)
    if kelvin < 0:
        raise ValueError("Temperature cannot be below 0 Kelvin")
    return kelvin - 273.15

# Conversion Configuration
# Dictionary mapping user choices to conversion functions and metadata
conversions = {
    '1': {
        'func': celsius_to_fahrenheit,
        'input_unit': 'Celsius',
        'output_unit': 'Fahrenheit',
        'prompt': 'Enter temperature in Celsius: '
    },
    '2': {
        'func': fahrenheit_to_celsius,
        'input_unit': 'Fahrenheit',
        'output_unit': 'Celsius',
        'prompt': 'Enter temperature in Fahrenheit: '
    },
    '3': {
        'func': celsius_to_kelvin,
        'input_unit': 'Celsius',
        'output_unit': 'Kelvin',
        'prompt': 'Enter temperature in Celsius: '
    },
    '4': {
        'func': kelvin_to_celsius,
        'input_unit': 'Kelvin',
        'output_unit': 'Celsius',
        'prompt': 'Enter temperature in Kelvin: '
    }
}

# User Interface Functions
def display_menu():
    # Displays the conversion menu to the user
    print("\nTemperature Converter")
    print("1. Celsius to Fahrenheit")
    print("2. Fahrenheit to Celsius")
    print("3. Celsius to Kelvin")
    print("4. Kelvin to Celsius")
    print("5. Exit")

def get_valid_choice():
    # Prompts the user for a menu choice and validates it
    # Returns a valid choice (1-5) or prompts again if invalid
    while True:
        choice = input("Enter your choice (1-5): ").strip()
        if choice in conversions or choice == '5':
            return choice
        print("Invalid choice! Please select between 1 and 5.")

def get_valid_temperature(prompt):
    # Prompts the user for a temperature value and ensures it's numeric
    # Returns a float or prompts again if invalid
    while True:
        try:
            return float(input(prompt))
        except ValueError:
            print("Invalid input! Please enter a numeric value.")

# Main Program
def main():
    # Main loop to run the temperature converter until the user chooses to exit
    while True:
        display_menu()  # Show the menu
        choice = get_valid_choice()  # Get a valid menu choice

        if choice == '5':  # Exit condition
            print("Exiting program.")
            break

        # Perform the selected conversion
        conversion = conversions[choice]
        try:
            temp = get_valid_temperature(conversion['prompt'])  # Get temperature input
            result = conversion['func'](temp)  # Apply conversion function
            # Display result formatted to 2 decimal places
            print(f"{temp:.2f}°{conversion['input_unit']} = {result:.2f}°{conversion['output_unit']}")
        except ValueError as e:
            print(f"Error: {e}")  # Handle invalid inputs or Kelvin < 0

#Entry Point
if __name__ == "__main__":
    # Ensures the program runs only when executed directly, not when imported
    main()

File: Java_Projects/Temperature_Converter.java
import java.util.Scanner;

public class Temperature_Converter {
    public static void main(String[] args) {
        // Create Scanner object to take user input
        Scanner scanner = new Scanner(System.in);
    
        System.out.println("===================================");
        System.out.println("   Welcome to Temperature Converter!");
        System.out.println("===================================");
        System.out.println();

        float temp = 0;  // Variable to store temperature value
        String unit = ""; // Variable to store unit type (celsius, fahrenheit, kelvin)

        // Input Temperature (with error handling)
        try {
            System.out.print("Enter Temperature: ");
            temp = scanner.nextFloat(); // Take numeric input for temperature
        } 
        catch (Exception e) {
            // Display message if invalid temperature entered
            System.out.println("Error: Enter a valid numeric temperature!");
            scanner.close();
            return; // Exit program on invalid input
        }

        scanner.nextLine(); 
        // Input Temperature Unit (with error handling)

        try {
            System.out.print("Enter Your Temperature Unit (celsius/kelvin/fahrenheit): ");
            unit = scanner.nextLine(); // Take unit input
            unit = unit.toLowerCase(); // Convert to lowercase to avoid case mismatch
        } 
        catch (Exception e) {
            // Display message if invalid unit entered
            System.out.println("Error: Enter Valid Unit!");
            scanner.close();
            return; // Exit if error occurs
        }

        switch (unit) {
            case "celsius":
                // Convert Celsius to Fahrenheit and Kelvin
                float f = (temp * 9 / 5) + 32;
                float k = temp + 273.15f;
                System.out.println(temp +" deg C = " + f +" deg F and " + k + "K");
                break;

            case "fahrenheit":
                // Convert Fahrenheit to Celsius and Kelvin
                float c1 = (temp - 32) * 5 / 9;
                float k1 = c1 + 273.15f;
                System.out.println(temp + " deg F = " + c1 + " deg C and " + k1 + "K");
                break;

            case "kelvin":
                // Convert Kelvin to Celsius and Fahrenheit
                float c2 = temp - 273.15f;
                float f2 = (c2 * 9 / 5) + 32;
                System.out.println(temp + "K = " + c2 + " deg C and " + f2 + " deg F");
                break;

            default:
                // Runs if user entered an invalid unit
                System.out.println("Invalid unit entered! Please enter celsius, fahrenheit, or kelvin only.");
                return;
        }

        // Close scanner to prevent memory leak
        scanner.close();
    }
}
